AlphaWallet/Tokens/Logic/SuggestEIP1559.swift

```diff
@@ -0,0 +1,209 @@
+                // In this case getting included with a low priority fee is not guaranteed; instead we use the higher base fee suggestion
+                // and also offer extra priority fee to increase the chance of getting included in the base fee dip.
+                guard let maxBaseFee = maxBaseFeeBi.optionalDecimalValue else { throw EIP1559Error.failureToBuildEstimates }
+
+                t = t.adding(maxBaseFee.subtracting(bf).multiplying(by: extraPriorityFeeRatio))
+                bfbi = maxBaseFeeBi
+            }
+            guard let tbi = BigInt(Int(truncating: t)) else { throw EIP1559Error.failureToBuildEstimates }
+
+            result[timeFactor] = EIP1559FeeOracleResult(maxFeePerGas: bfbi + tbi, maxPriorityFeePerGas: tbi, baseFee: bfbi)
+        }
+
+        return result
+    }
+
+    private func predictMinBaseFee(baseFee: [BigInt], order: [Int], timeFactor: Double) -> BigInt {
+        assert(!baseFee.isEmpty)
+
+        if timeFactor < 1e-6 {
+            return baseFee[baseFee.count - 1]
+        }
+
+        let pendingWeight = (1 - exp(-1 / timeFactor)) / (1 - exp(-Double(baseFee.count) / timeFactor))
+        var sumWeight: Double = 0.0
+        var result: BigInt = .zero
+        var samplingCurveLast: NSDecimalNumber = 0.0
+
+        for i in order.indices {
+            sumWeight += pendingWeight * exp(Double((order[i] - baseFee.count + 1)) / timeFactor)
+            let samplingCurveValue = samplingCurve(percentile: sumWeight * 100.0)
+            guard let indexedFee = baseFee[order[i]].optionalDecimalValue else { fatalError() }
+            let value = samplingCurveValue.subtracting(samplingCurveLast).multiplying(by: indexedFee)
+            guard let valueBi = BigInt(Int(truncating: value)) else { fatalError() }
+
+            result += valueBi
+
+            if samplingCurveValue.doubleValue >= 1 {
+                return result
+            }
+            samplingCurveLast = samplingCurveValue
+        }
+
+        return result
+    }
+
+    private func samplingCurve(percentile: Double) -> NSDecimalNumber {
+        if percentile <= sampleMinPercentile {
+            return 0.0
+        } else if percentile >= sampleMaxPercentile {
+            return 1.0
+        } else {
+            return NSDecimalNumber(value: Double((1 - cos((percentile - sampleMinPercentile) * 2 * Double.pi / (sampleMaxPercentile - sampleMinPercentile))) / 2))
+        }
+    }
+
+    private func maxBlockCount(gasUsedRatio: [Double], _ _ptr: Int, _ _needBlocks: Int) -> Int {
+        var blockCount = 0
+        var ptr = _ptr
+        var needBlocks = _needBlocks
+        while needBlocks > 0 && ptr >= 0 {
+            if gasUsedRatio[ptr] == 0.0 || gasUsedRatio[ptr] > 0.9 {
+                break
+            }
+            ptr -= 1
+            needBlocks -= 1
+            blockCount += 1
+        }
+        return blockCount
+    }
+}
+
+public struct EIP1559FeeOracleResult {
```

Comment:
Probably should rename this as not only used for "oracle" result

Reply:
used the same name as in android impl